src/compiler/crystal/command.cr

```diff
@@ -521,17 +520,29 @@ class Crystal::Command
     if has_stdin_filename
       sources << Compiler::Source.new(filenames.shift, STDIN.gets_to_end)
     end
-    sources += gather_sources(filenames)
-    first_filename = sources.first.filename
-    first_file_ext = File.extname(first_filename)
-    original_output_filename = File.basename(first_filename, first_file_ext)
+    sources.concat gather_sources(filenames)
 
-    # Check if we'll overwrite the main source file
-    if first_file_ext.empty? && !output_filename && !no_codegen && !run && first_filename == File.expand_path(original_output_filename)
-      error "compilation will overwrite source file '#{Crystal.relative_filename(first_filename)}', either change its extension to '.cr' or specify an output file with '-o'"
+    if output_filename
+      if compiler.codegen_target.win32?
+        # foo -> foo.exe
+        # foo.bar -> foo.exe
+        # foo.exe -> foo.exe
+        # foo.tar.gz -> foo.tar.exe
+        output_filename = "#{output_filename.rchop(File.extname(output_filename))}.exe"
+      end
+    else
+      first_filename = sources.first.filename
+      output_filename = File.basename(first_filename, File.extname(first_filename))
```

Comment:
This is equivalent:
```suggestion
      output_filename = Path[first_filename].stem
```

(`Path#stem` is currently not exposed as a class method on `File`.)